Split `BoundInterpolatedString` into `BoundInterpolatedString` and `BoundUnconvertedInterpolatedString`

333fred opened the pull request:
This is intended to be a semantics-preserving refactoring that will make later work easier. When an interpolated string is first encountered, we now bind it as a `BoundUnconvertedInterpolatedString` node, and then when we bind to the natural type we convert it to a `BoundInterpolatedString` expression. Future code to allow it to convert to non-string types will take advantage of this in the conversion step.

333fred commented:
@chsienki for review.

chsienki commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -374,6 +374,18 @@ internal BoundExpression BindToNaturalType(BoundExpression expression, BindingDi
+                    {
+                        // We determine the best method of emitting as a string (either via Concat, Format, or the builder pattern)
+                        // during lowering, and it's not part of the publicly-visible API, unlike conversion to a builder type.
+                        result = new BoundInterpolatedString(
```
I'm not sure I follow this comment. If we do building in lowering, why do we need to have this split here at all? Or am I confusing builder pattern and builder type as the same thing?

My assumption is that we'll add code here in  future PR to create some new kind of `BoundBuilderInterpoltedString` if we detect its actually a builder pattern/type one right?

333fred replied:
The code won't be here. The code will be when the interpolated string is converted to a non-`String` type.

chsienki replied:
Hmm, I guess I'm missing something. Why do we need the `BoundUnconvertedInterpolatedString` / `BoundInterpolatedString` if we're converting to another type later on?

333fred replied:
> Why do we need the BoundUnconvertedInterpolatedString / BoundInterpolatedString if we're converting to another type later on?

We need it because we want to tell the difference between a string that has been converted and one that has not been converted. This is the same approach we take with other target-typed constructs such as switch expressions and ternaries. Happy to discuss over teams in more depth if you want.

chsienki replied:
Ahh, got it, thanks.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundExpression.cs:
```diff
@@ -286,6 +287,14 @@ internal partial class BoundBinaryOperator
         }
     }
 
+    internal partial class BoundUnconvertedInterpolatedString
```
>BoundUnconvertedInterpolatedString [](start = 27, length = 34)

Can we override the method once in BoundInterpolatedStringBase?

chsienki commented on src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs:
```diff
@@ -299,7 +299,7 @@ private ExpressionSyntax SkipParensAndNullSuppressions(ExpressionSyntax e)
                 inputType, patternExpression, expression, out constantValueOpt, hasErrors, diagnostics);
 
-            ConstantValueUtils.CheckLangVersionForConstantValue(expression, diagnostics);
+            ConstantValueUtils.CheckLangVersionForConstantValue(convertedExpression, diagnostics);
```
This is an unrelated bug?

333fred replied:
> This is an unrelated bug? Or just expressing itself now?

Neither. Previously, `BoundInterpolatedString` was never converted, so `expression` and `convertedExpression` were the same node.

AlekseyTs commented:
>                 ? Conversion.InterpolatedString : Conversion.NoConversion;

Should these conversions be allowed only from an unconverted form?

---
Refers to: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs:87 in 01de52a. [](commit_id = 01de52a2efdd11c2dbe7a0bcda2d1554c98afc90, deletion_comment = False)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs:
```diff
@@ -72,7 +72,7 @@ internal ConversionsBase WithNullability(bool includeNullability)
         protected abstract ConversionsBase CreateInstance(int currentRecursionDepth);
 
-        protected abstract Conversion GetInterpolatedStringConversion(BoundInterpolatedString source, TypeSymbol destination, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo);
+        protected abstract Conversion GetInterpolatedStringConversion(BoundInterpolatedStringBase source, TypeSymbol destination, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo);
```
>BoundInterpolatedStringBase [](start = 70, length = 27)

It feels like this should be BoundUnconvertedInterpolatedString instead.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:
```diff
@@ -1056,6 +1056,11 @@ public override BoundNode VisitInterpolatedString(BoundInterpolatedString node)
             return null;
         }
 
+        public override BoundNode VisitInterpolatedString(BoundInterpolatedString node)
```
Are we not expecting to run into an uconverted node? I think we should still override the visit method and perhaps throw.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -9362,6 +9362,14 @@ private static bool IsNullabilityMismatch(TypeSymbol type1, TypeSymbol type2)
+        public override BoundNode? VisitUnconvertedInterpolatedString(BoundUnconvertedInterpolatedString node)
+        {
+            // This is only involved with unbound lambdas or when visiting the source of a converted tuple literal
+            var result = base.VisitInterpolatedStringBase(node);
```
>VisitInterpolatedStringBase [](start = 30, length = 27)

Why not VisitUnconvertedInterpolatedString?

333fred replied:
With your other suggestion, this would now throw, so I'll leave this as is.

---
In reply to: [601832232](https://github.com/dotnet/roslyn/pull/52061#discussion_r601832232) [](ancestors = 601832232)

333fred replied:
Updated based on our offline discussion.

---
In reply to: [601847314](https://github.com/dotnet/roslyn/pull/52061#discussion_r601847314) [](ancestors = 601847314,601832232)

AlekseyTs commented:
Done with review pass (commit 1)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [807471528](https://github.com/dotnet/roslyn/pull/52061#issuecomment-807471528) [](ancestors = 807471528)

AlekseyTs reviewed (approved):
LGTM (commit 3)
